python/pyspark/rdd.py

```diff
@@ -2349,7 +2352,12 @@ def toLocalIterator(self):
         """
         with SCCallSiteSync(self.context) as css:
             port = self.ctx._jvm.PythonRDD.toLocalIteratorAndServe(self._jrdd.rdd())
-        return _load_from_socket(port, self._jrdd_deserializer)
+        # We set a timeout for connecting socket. The connection only begins when we start
+        # to consume the first element. If we do not begin to consume the returned iterator
+        # immediately, there will be a failure.
+        iter = _load_from_socket(port, self._jrdd_deserializer)
```

Comment:
I think we could have a wrapper to load items, and return the iterator in _load_from_socket(): 
```
def _load_from_socket(port):
     ...
    return  serializer.load_stream(sock.makefile("rb", 65536))
```

The `finally` may be optional, GC cloud close the socket (should be confirmed).